### PR TITLE
Support pyspark.sql.types.ArrayType with pyspark.sql.types.StringType elements

### PR DESCRIPTION
**Pull request by nyamashi**
This modification has been tested using  ```pyzoo/test/zoo/orca/learn/ray/tf/test_tf_ray_estimator.py``` adding ```test_array_string_input()```.   (I also modified ```test_string_input()```  to adapt a vocabulary.)

results are:
```test_array_string_input()```:
```
[Row(id=0, input=['foo', 'qux', 'bar'], prediction=DenseVector([4.0, 1.0, 2.0])), Row(id=1, input=['qux', 'baz'], prediction=DenseVector([1.0, 3.0]))]
```
```test_string_input()```:
```
[Row(input='foo qux bar', prediction=DenseVector([3.0, 2.0, 5.0, 0.0])), Row(input='qux baz', prediction=DenseVector([2.0, 4.0, 0.0, 0.0]))]
```

test env:
- python3.7
- tensorflow==2.7.0
- ray==1.9.2

### CODE REVIEW

**Review by hkvision — APPROVED** [review]
Thanks for your contribution!

**Comment by nyamashi** [comment]
@hkvision  Thank you for your support!!

 I had forgotten to check the TF version when I wrote the test code, sorry.

The test code I added is based on variable length data with the TF ragged tensor because I'm working on such ML problems.
It can be run but I found there are many difficulties to use variable length data fields for more general models with Orca APIs. 
(i.e ```np.stack``` requires a argument to be the same shape/length: https://github.com/intel-analytics/analytics-zoo/blob/0aad45f44e9b4f35ed746f3bd0f44c232f2dbeb9/pyzoo/zoo/orca/learn/utils.py#L210 )
So it would be better to use fixed length data and model for a test as like ```test_string_input()```.. 🤔

**Comment by hkvision** [comment]
> test_string_input

Yeah, at this moment, we assume the arrays have fixed length so that we can combine them to batches and if they are of variable length, they should be padded first.
If you have some available sample code for variable length data, it will be great to share with us so that we can think of better supporting such case in Orca :)
Thank you so much!

**Comment by hkvision** [comment]
@nyamashi Unfortunately tensorflow in our environment is 2.3.0,
could you kindly change the import to `tf.keras.layers.experimental.preprocessing` for both TextVectorization and StringLookup?

Thanks so much :)

**Comment by nyamashi** [comment]
> Yeah, at this moment, we assume the arrays have fixed length so that we can combine them to batches and if they are of variable length, they should be padded first.

I understand that padding is a standard method but I tried a shortcut at first(always I try and mostly fail). I'll switch to the plan-B.

> If you have some available sample code for variable length data, it will be great to share with us so that we can think of better supporting such case in Orca :)

There is no sample code that I can share now, but I'll prepare a sample using public dataset like MovieLens. 
I'm working around recommender systems and just trying to adapt https://www.tensorflow.org/recommenders  to PySpark/Ray/Orca environment.  This PR originally comes from https://www.tensorflow.org/recommenders/examples/listwise_ranking.

Thanks!

**Comment by nyamashi** [comment]
@hkvision 
I've changed the environment with tensorflow==2.3.0 and reverted  ```test_string_input()```.
I also modified ```test_array_string_input()``` with fixed length input data. 
Please check it.

results are:
```test_array_string_input():```
````
[Row(id=0, input=['foo', 'qux', 'bar'], prediction=DenseVector([2.0, 5.0, 4.0])), Row(id=1, input=['qux', 'baz', 'baz'], prediction=DenseVector([5.0, 3.0, 3.0]))]
````
and ```test_string_input():```
```
[Row(input='foo qux bar', prediction=DenseVector([1.0, 1.0, 1.0, 0.0])), Row(input='qux baz', prediction=DenseVector([1.0, 1.0, 0.0, 0.0]))]
```

**Comment by hkvision** [comment]
> 

@nyamashi Thanks! The changes look good :)

PEP8 checks failed.
./test/zoo/orca/learn/ray/tf/test_tf_ray_estimator.py:639:101: E501 line too long (104 > 100 characters)

One style issue there, could you make Line 639 into two separate lines so that I can merge your PR?
BTW, you can run `pyzoo/dev/lint-python` for the style check on your machine :)

**Comment by hkvision** [comment]
> > Yeah, at this moment, we assume the arrays have fixed length so that we can combine them to batches and if they are of variable length, they should be padded first.
> 
> I understand that padding is a standard method but I tried a shortcut at first(always I try and mostly fail). I'll switch to the plan-B.
> 
> > If you have some available sample code for variable length data, it will be great to share with us so that we can think of better supporting such case in Orca :)
> 
> There is no sample code that I can share now, but I'll prepare a sample using public dataset like MovieLens. I'm working around recommender systems and just trying to adapt https://www.tensorflow.org/recommenders to PySpark/Ray/Orca environment. This PR originally comes from using Orca for https://www.tensorflow.org/recommenders/examples/listwise_ranking.
> 
> Thanks!

Thank you for the information provided. Actually we are also doing some work for recommendation in BigDL (analytics-zoo has been merged into BigDL and we are now developing in BigDL): https://github.com/intel-analytics/BigDL/tree/branch-2.0/python/friesian/example
Feel free to tell raise issues if you encounter questions in Orca later and we are very glad to help. If there's some progress in listwise ranking that you can make them public, it will be great to raise a PR to our project!

**Comment by nyamashi** [comment]
@hkvision 
I didn't realize the linter ```pyzoo/dev/lint-python``` and thank you for letting me know.

**Comment by hkvision** [comment]
http://10.112.231.51:18888/job/ZOO-PR-Validation/6538/ The failures are unrelated to this PR. http://10.112.231.51:18888/job/ZOO-PR-Python-Spark-3.1-py37-Horovod/1992/console The newly added unit test passed.

**Comment by hkvision** [comment]
Hi @nyamashi 

I've merged this PR. Thank you so much for your contribution!
BTW, as we have merged analytics-zoo to BigDL branch 2.0, would you mind opening a PR for these changes in BigDL here? https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/orca/test/bigdl/orca/learn/ray/tf/test_tf_ray_estimator.py
Or I can do this for you :)

Thanks again!

**Comment by nyamashi** [comment]
@hkvision

Okay I'll do it!
I was just watching only https://github.com/intel-analytics/analytics-zoo and haven't checked https://github.com/intel-analytics/BigDL somehow. Maybe I misunderstand the relation between Analytics-Zoo and BigDL..

Thanks!!

**Comment by hkvision** [comment]
> @hkvision
> 
> Okay I'll do it! I was just watching only https://github.com/intel-analytics/analytics-zoo and haven't checked https://github.com/intel-analytics/BigDL somehow. Maybe I misunderstand the relation between Analytics-Zoo and BigDL..
> 
> Thanks!!

Thanks! 
We have combined the analytics-zoo and original BigDL project into the version 2.0 of BigDL :)

**Comment by nyamashi** [comment]
Hi @hkvision 

I tried to make a BigDL pip package before adding this modification, 
(It would be https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/dllib/src/bigdl/dllib/utils/utils.py and  https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/orca/test/bigdl/orca/learn/ray/tf/test_tf_ray_estimator.py
)
but I couldn't make it for some reason.
One of the errors I got is like:
```
$ bash ./python/dev/release_default_linux_spark246.sh 0.14.0 false false false
...
$ pip install python/dist/bigdl-0.14.0-py3-none-manylinux1_x86_64.whl      
Processing ./python/dist/bigdl-0.14.0-py3-none-manylinux1_x86_64.whl
ERROR: Could not find a version that satisfies the requirement bigdl-chronos==0.14.0 (from bigdl) (from versions: 0.14.0b20211020, 0.14.0b20211022, 0.14.0b20211025, 0.14.0b20211026, 0.14.0b20211027, 0.14.0b20211028, 0.14.0b20211101, 0.14.0b20211102, 0.14.0b20211103, 0.14.0b20211104, 0.14.0b20211105, 0.14.0b20211106, 0.14.0b20211107, 0.14.0b20211108, 0.14.0b20211109, 0.14.0b20211110, 0.14.0b20211111, 0.14.0b20211203, 0.14.0b20211204, 0.14.0b20211205, 0.14.0b20211206, 0.14.0b20211207, 0.14.0b20211208, 0.14.0b20211209, 0.14.0b20211210, 0.14.0b20211211, 0.14.0b20211212, 0.14.0b20211213, 0.14.0b20211214, 0.14.0b20211215, 0.14.0b20220107, 0.14.0b20220108, 0.14.0b20220109, 0.14.0b20220110, 0.14.0b20220111, 0.14.0b20220112, 0.14.0b20220113, 0.14.0b20220114, 0.14.0b20220115, 0.14.0b20220116, 0.14.0b20220117, 0.14.0b20220118, 0.14.0b20220120, 0.14.0b20220121, 0.14.0b20220122, 0.14.0b20220124, 0.14.0b20220126, 0.14.0b20220128, 0.14.0b20220201, 0.14.0b20220203, 0.14.0b20220205, 0.14.0b20220206, 0.14.0b20220208, 0.14.0b20220209, 0.14.0b20220210, 0.14.0b20220211, 0.14.0b20220212, 0.14.0b20220213)
ERROR: No matching distribution found for bigdl-chronos==0.14.0
```
In analytics-zoo case, I referred to https://analytics-zoo.readthedocs.io/en/latest/doc/UserGuide/develop.html
(The BigDL's guide seems same as analytics-zoo still: https://bigdl.readthedocs.io/en/latest/doc/UserGuide/develop.html 😄 )
Could you tell me the way to build a BigDL package to check this modification?

Thank you!

**Comment by hkvision** [comment]
Hi @nyamashi 

For BigDL, we have adopted modular designs and each submodule would have it's own package.
After you run the build script, you need to check python/dllib|orca/src/dist and install the wheel in that folders. (dllib is the original BigDL and Orca is migrated from analytics-zoo).
You need to first install the dllib whl and followed by orca whl.

Sorry that our document hasn't been updated and there's an issue to track this: #3995

Feel free to tell me if you have any questions for BigDL installation and usage.

**Comment by nyamashi** [comment]
@hkvision  
Thanks for your comments!
I understood the BigDL repo designs and finally I could make a packages.
I checked the same modification in the BigDL repo, so I could made a PR. https://github.com/intel-analytics/BigDL/pull/4012 

Thank you for you support again!
